### PR TITLE
19834 convert streamutils graphstage

### PR DESCRIPTION
**Pull request by leachbj**

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

File: akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala [hunk 1]
```diff
@@ -50,15 +51,17 @@ private[http] object StreamUtils {
...
+        override def onPull(): Unit = pull(in)
 
-      override def onUpstreamFailure(cause: Throwable, ctx: Context[ByteString]): TerminationDirective =
-        ctx.fail(f(cause))
+        setHandlers(in, out, this)
+      }
     }
 
-    Flow[ByteString].transform(() ⇒ transformer).named("transformError")
+    Flow[ByteString].via(transformer).named("transformError")
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala]: could be replaced with `.recover { t => throw f(t)` I think?

**Comment by ktoso** [comment]
Thanks, looks good - LGTM

File: akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala (hunk 1 above)
Comment (leachbj) [inline akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala]: I just went with a literal translation, probably worth looking at a refactor of the callers though;

`CodingDirectives.decodeRequestWith` looks like the only user.

    effectiveDecoder.decode(request).mapEntity(StreamUtils.mapEntityError {
              case NonFatal(e) ⇒
                IllegalRequestException(
                  StatusCodes.BadRequest,
                  ErrorInfo("The request's encoding is corrupt", e.getMessage))
            })

Perhaps just refactor to 

    effectiveDecoder.decode(request).mapEntity { entity ⇒
              entity.transformDataBytes(Flow[ByteString].recover {
                case NonFatal(e) ⇒
                  throw IllegalRequestException(
                    StatusCodes.BadRequest,
                    ErrorInfo("The request's encoding is corrupt", e.getMessage))
              })
            }```

Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala]: Yes, this looks good to me – would you apply this change here?
In general we're trying to get rid of the special impls inside of StreamUtils :)

**Comment by ktoso** [comment]
Excellent, thanks!

LGTM

**Comment by drewhk** [comment]
LGTM

**Comment by drewhk** [comment]
It only needs the commits to be squashed and then it can be merged

**Comment by ktoso** [comment]
We can do the squashing now btw @drewhk https://github.com/blog/2141-squash-your-commits :-)

**Comment by drewhk** [comment]
totally awesome!

**Pull request merged**